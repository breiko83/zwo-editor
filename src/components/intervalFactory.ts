import { v4 as uuidv4 } from 'uuid'
import { PaceType } from './Editor/PaceSelector';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "./Interval";

const defaultDuration = 300;
const defaultPower = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

export default {
  steady(interval: Partial<SteadyInterval>): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      duration: defaultDuration,
      power: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      duration: defaultDuration,
      startPower: defaultPower / 2,
      endPower: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      duration: defaultDuration,
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      cadence: defaultCadence,
      restingCadence: defaultCadence,
      repeat: 3,
      onDuration: 30,
      offDuration: 120,
      onPower: defaultPower,
      offPower: defaultPower / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
