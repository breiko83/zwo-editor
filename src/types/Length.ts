export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;
