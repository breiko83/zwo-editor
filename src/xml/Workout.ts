import { SportType } from "../components/Editor/Editor";
import { Instruction } from "../components/Instruction";
import { Interval } from "../components/Interval";

export interface Workout {
  author: string;
  name: string;
  description: string;
  sportType: SportType;
  tags: string[];
  intervals: Array<Interval>;
  instructions: Array<Instruction>;
}

export function createEmptyWorkout(): Workout {
  return {
    author: "",
    name: "",
    description: "",
    sportType: "bike",
    tags: [],
    intervals: [],
    instructions: [],
  };
}
