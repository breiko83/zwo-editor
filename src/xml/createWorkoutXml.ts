import Builder from 'xmlbuilder'
import RunMode from '../modes/RunMode';
import { WorkoutMode } from '../modes/WorkoutMode';
import { Instruction } from '../types/Instruction';
import { Interval } from '../types/Interval';
import { Duration, Length } from '../types/Length'
import { Workout } from '../types/Workout'

export default function createWorkoutXml({ author, name, description, sportType, lengthType, tags, intervals, instructions }: Workout, mode: WorkoutMode): string {
  var totalLength = 0;

  let xml = Builder.begin()
    .ele('workout_file')
    .ele('author', author).up()
    .ele('name', name).up()
    .ele('description', description).up()
    .ele('sportType', sportType).up()
    .ele('durationType', lengthType).up()
    .ele('tags')

  tags.forEach((tag: string) => {
    var t: Builder.XMLNode
    t = Builder.create('tag')
      .att('name', tag)
    xml.importDocument(t)
  })

  xml = xml.up().ele('workout')

  const writeLength = (len: Length): number =>
    len instanceof Duration ? len.seconds : len.meters;

  intervals.forEach((interval, index) => {
    var segment: Builder.XMLNode
    var ramp

    if (interval.type === 'steady') {
      segment = Builder.create('SteadyState')
        .att('Duration', writeLength(interval.length))
        .att('Power', interval.intensity)
        .att('pace', interval.pace)

      // add cadence if not zero
      interval.cadence !== 0 && segment.att('Cadence', interval.cadence)

    } else if (interval.type === 'ramp' && interval.startIntensity && interval.endIntensity) {

      // index 0 is warmup
      // last index is cooldown
      // everything else is ramp

      ramp = 'Ramp'
      if (index === 0) ramp = 'Warmup'
      if (index === intervals.length - 1) ramp = 'Cooldown'

      if (interval.startIntensity < interval.endIntensity) {
        // warmup
        segment = Builder.create(ramp)
          .att('Duration', writeLength(interval.length))
          .att('PowerLow', interval.startIntensity)
          .att('PowerHigh', interval.endIntensity)
          .att('pace', interval.pace)
        // add cadence if not zero
        interval.cadence !== 0 && segment.att('Cadence', interval.cadence)
      } else {
        // cooldown
        segment = Builder.create(ramp)
          .att('Duration', writeLength(interval.length))
          .att('PowerLow', interval.startIntensity) // these 2 values are inverted
          .att('PowerHigh', interval.endIntensity) // looks like a bug on zwift editor            
          .att('pace', interval.pace)
        // add cadence if not zero
        interval.cadence !== 0 && segment.att('Cadence', interval.cadence)
      }
    } else if (interval.type === 'repetition') {
      // <IntervalsT Repeat="5" OnDuration="60" OffDuration="300" OnPower="0.8844353" OffPower="0.51775455" pace="0"/>
      segment = Builder.create('IntervalsT')
        .att('Repeat', interval.repeat)
        .att('OnDuration', writeLength(interval.onLength))
        .att('OffDuration', writeLength(interval.offLength))
        .att('OnPower', interval.onIntensity)
        .att('OffPower', interval.offIntensity)
        .att('pace', interval.pace)
        // add cadence if not zero
        interval.onCadence !== 0 && segment.att('Cadence', interval.onCadence)
        // add cadence resting if not zero
        interval.offCadence !== 0 && segment.att('CadenceResting', interval.offCadence)
    } else {
      // free ride
      segment = Builder.create('free')
        .att('Duration', writeLength(interval.length))
        .att('FlatRoad', 0) // Not sure what this is for
      // add cadence if not zero
      interval.cadence !== 0 && segment.att('Cadence', interval.cadence)
    }

    const intervalLength = (interval: Interval): number =>
      mode instanceof RunMode && mode.lengthType === "distance"
        ? mode.intervalDistance(interval).meters
        : mode.intervalDuration(interval).seconds;

    const instructionInsideInterval = (instruction: Instruction): boolean =>
      (writeLength(instruction.offset) >= totalLength && writeLength(instruction.offset) < (totalLength + intervalLength(interval)));

    // add instructions if present
    instructions.filter(instructionInsideInterval).forEach((i) => {
      segment.ele('textevent', { timeoffset: (writeLength(i.offset) - totalLength), message: i.text })
    })

    xml.importDocument(segment)

    totalLength += intervalLength(interval);
  })

  return xml.end({ pretty: true });
}
