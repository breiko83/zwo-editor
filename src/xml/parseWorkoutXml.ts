import Converter from 'xml-js';
import { v4 as uuidv4 } from 'uuid'
import intervalFactory from '../interval/intervalFactory';
import { createEmptyWorkout, Workout } from '../types/Workout';
import { Distance, Duration } from '../types/Length';
import { WorkoutMode } from '../modes/WorkoutMode';

export default function parseWorkoutXml(data: string, mode: WorkoutMode): Workout {
  const workout: Workout = createEmptyWorkout(mode.sportType, mode.lengthType);

  data = data.replace(/<!--(.*?)-->/gm, "")

  const workout_file = Converter.xml2js(data).elements[0]

  if (workout_file.name !== 'workout_file') {
    throw new Error("Invalid ZWO file");
  }

  const sportTypeIndex = workout_file.elements.findIndex((element: { name: string }) => element.name === 'sportType')
  if (sportTypeIndex !== -1 && workout_file.elements[sportTypeIndex].elements) {
    workout.sportType = workout_file.elements[sportTypeIndex].elements[0].text
  }

  const lengthTypeIndex = workout_file.elements.findIndex((element: { name: string }) => element.name === 'durationType')
  if (lengthTypeIndex !== -1 && workout_file.elements[lengthTypeIndex].elements) {
    workout.lengthType = workout_file.elements[lengthTypeIndex].elements[0].text
  }

  const authorIndex = workout_file.elements.findIndex((element: { name: string }) => element.name === 'author')
  if (authorIndex !== -1 && workout_file.elements[authorIndex].elements) {
    workout.author = workout_file.elements[authorIndex].elements[0].text
  }

  const nameIndex = workout_file.elements.findIndex((element: { name: string }) => element.name === 'name')
  if (nameIndex !== -1 && workout_file.elements[nameIndex].elements) {
    workout.name = workout_file.elements[nameIndex].elements[0].text
  }

  const descriptionIndex = workout_file.elements.findIndex((element: { name: string }) => element.name === 'description')
  if (descriptionIndex !== -1 && workout_file.elements[descriptionIndex].elements) {
    workout.description = workout_file.elements[descriptionIndex].elements[0].text;
  }

  const tagsIndex = workout_file.elements.findIndex((element: { name: string }) => element.name === 'tags')
  if (tagsIndex !== -1 && workout_file.elements[tagsIndex].elements) {
    workout.tags = workout_file.elements[tagsIndex].elements.map(
      (el: { attributes: { name: string } }) => el.attributes.name
    );
  }

  const workoutEl = workout_file.elements.find((element: { name: string }) => element.name === 'workout' && "elements" in element)
  if (!workoutEl) {
    return workout;
  }

  // either meters (distance) or seconds (duration)
  let totalLength = 0

  const readLength = (x: number) =>
    workout.lengthType === "time" ? new Duration(x) : new Distance(x);

  workoutEl.elements.map((w: { name: string; attributes: { Power: any; PowerLow: string; Duration: string; PowerHigh: string; Cadence: string; CadenceResting: string; Repeat: string; OnDuration: string; OffDuration: string; OnPower: string, OffPower: string; pace: string }; elements: any }) => {

    let length = parseFloat(w.attributes.Duration)

    if (w.name === 'SteadyState') {
      workout.intervals.push(intervalFactory.steady({
        intensity: parseFloat(w.attributes.Power || w.attributes.PowerLow),
        length: readLength(parseFloat(w.attributes.Duration)),
        cadence: parseFloat(w.attributes.Cadence || '0'),
        pace: parseInt(w.attributes.pace || '0'),
      }, mode))
    }
    if (w.name === 'Ramp' || w.name === 'Warmup' || w.name === 'Cooldown') {
      workout.intervals.push(intervalFactory.ramp({
        startIntensity: parseFloat(w.attributes.PowerLow),
        endIntensity: parseFloat(w.attributes.PowerHigh),
        length: readLength(parseFloat(w.attributes.Duration)),
        pace: parseInt(w.attributes.pace || '0'),
        cadence: parseInt(w.attributes.Cadence || '0'),
      }, mode))
    }
    if (w.name === 'IntervalsT') {
      workout.intervals.push(intervalFactory.repetition({
        repeat: parseFloat(w.attributes.Repeat),
        onLength: readLength(parseFloat(w.attributes.OnDuration)),
        offLength: readLength(parseFloat(w.attributes.OffDuration)),
        onIntensity: parseFloat(w.attributes.OnPower),
        offIntensity: parseFloat(w.attributes.OffPower),
        onCadence: parseInt(w.attributes.Cadence || '0'),
        offCadence: parseInt(w.attributes.CadenceResting || '0'),
        pace: parseInt(w.attributes.pace || '0'),
      }, mode))
      length = (parseFloat(w.attributes.OnDuration) + parseFloat(w.attributes.OffDuration)) * parseFloat(w.attributes.Repeat)
    }
    if (w.name === 'free') {
      workout.intervals.push(intervalFactory.free({
        length: readLength(parseFloat(w.attributes.Duration)),
        cadence: parseInt(w.attributes.Cadence || '0'),
      }, mode))
    }

    // check for instructions
    const textElements = w.elements
    if (textElements && textElements.length > 0) {
      textElements.forEach((t: { name: string; attributes: { message: string | undefined; timeoffset: string } }) => {
        if (t.name.toLowerCase() === 'textevent')
          workout.instructions.push({
            text: t.attributes.message || '',
            offset: readLength(totalLength + parseFloat(t.attributes.timeoffset)),
            id: uuidv4()
          });
      })
    }

    totalLength = totalLength + length
    // map functions expect return value
    return false
  })

  return workout;
}
