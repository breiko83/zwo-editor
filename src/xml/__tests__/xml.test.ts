import intervalFactory from "../../interval/intervalFactory";
import createMode from "../../modes/createMode";
import { createInstruction } from "../../types/Instruction";
import { Distance, Duration } from "../../types/Length";
import { PaceType } from "../../types/PaceType";
import { Workout } from "../../types/Workout";
import createWorkoutXml from "../createWorkoutXml";
import parseWorkoutXml from "../parseWorkoutXml";

jest.mock('uuid', () => ({
  v4: () => "mock-id",
}));

describe("XML", () => {
  it('creates and parses an empty workout', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});
    const workout: Workout = {
      name: "Sample Workout",
      author: "John Doe",
      description: "Some intervals for your pleasure (or pain).",
      tags: ["RECOVERY", "INTERVALS"],
      sportType: mode.sportType,
      lengthType: mode.lengthType,
      intervals: [],
      instructions: [],
    };
    const xml = createWorkoutXml(workout, mode);
    expect(xml).toMatchSnapshot();

    // Parsing of tags not yet supported
    expect(parseWorkoutXml(xml, mode)).toEqual(workout);
  });

  it('creates and parses bike workout (without cadence)', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});
    const workout: Workout = {
      name: "Cycling Workout",
      author: "Eddy Merckx",
      description: "Featuring a mixture of different interval types.",
      tags: [],
      sportType: mode.sportType,
      lengthType: mode.lengthType,
      intervals: [
        // Warmup: 10:00 30%..75%
        intervalFactory.ramp({ length: new Duration(10 * 60), startIntensity: 0.3, endIntensity: 0.75 }, mode),
        // Steady: 5:00 80%
        intervalFactory.steady({ length: new Duration(5 * 60), intensity: 0.8 }, mode),
        // Intervals: 4 x ON 0:30 120%, OFF 1:00 60%
        intervalFactory.repetition({ repeat: 4, onLength: new Duration(30), onIntensity: 1.2, offLength: new Duration(60), offIntensity: 0.6 }, mode),
        // Ramp 1:00 80%..150%
        intervalFactory.ramp({ length: new Duration(60), startIntensity: 0.8, endIntensity: 1.5 }, mode),
        // Freeride: 20:00
        intervalFactory.free({ length: new Duration(20 * 60) }, mode),
        // Cooldown: 10:00 30%..30%
        intervalFactory.ramp({ length: new Duration(10 * 60), startIntensity: 0.7, endIntensity: 0.3 }, mode),
      ],
      instructions: [
        // @ 00:00 - at the start of warmup
        createInstruction({ offset: new Duration(0), text: "Welcome to the workout!" }, mode),
        // @ 12:30 - at the middle of second block
        createInstruction({ offset: new Duration(12 * 60 + 30), text: "This is just a warmup still" }, mode),
        // @ 15:00 - at the start of repetition block
        createInstruction({ offset: new Duration(15 * 60), text: "It's the first one of four sprint efforts" }, mode),
        // @ 19:30 - at the start of last repetition block
        createInstruction({ offset: new Duration(19 * 60 + 30), text: "It's the first one of four sprint efforts" }, mode),
        // @ 21:00 - at the start of a ramp
        createInstruction({ offset: new Duration(21 * 60), text: "As a bonus, we'll ramp up really hard :)" }, mode),
        // @ 32:00 - inside freeride
        createInstruction({ offset: new Duration(32 * 60), text: "Ride as hard as you can for 20 minutes!" }, mode),
        // @ 51:50 - almost at the very end of cooldown
        createInstruction({ offset: new Duration(51 * 60 + 50), text: "This was it. See you next time." }, mode),
      ],
    };
    const xml = createWorkoutXml(workout, mode);
    expect(xml).toMatchSnapshot();

    expect(parseWorkoutXml(xml, mode)).toEqual(workout);
  });

  it('creates and parses bike workout (with cadence)', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});
    const workout: Workout = {
      name: "Cycling Workout",
      author: "Eddy Merckx",
      description: "Featuring a mixture of different interval types.",
      tags: [],
      sportType: mode.sportType,
      lengthType: mode.lengthType,
      intervals: [
        // Warmup: 10:00 30%..75% 80rpm
        intervalFactory.ramp({ length: new Duration(10 * 60), startIntensity: 0.3, endIntensity: 0.75, cadence: 80 }, mode),
        // Steady: 5:00 80% 90rpm
        intervalFactory.steady({ length: new Duration(5 * 60), intensity: 0.8, cadence: 90 }, mode),
        // Intervals: 4 x ON 0:30 120% 100rpm, OFF 1:00 60% 60rpm
        intervalFactory.repetition({ repeat: 4, onLength: new Duration(30), onIntensity: 1.2, onCadence: 100, offLength: new Duration(60), offIntensity: 0.6, offCadence: 60 }, mode),
        // Ramp 1:00 80%..150% 90rpm
        intervalFactory.ramp({ length: new Duration(60), startIntensity: 0.8, endIntensity: 1.5, cadence: 90 }, mode),
        // Freeride: 20:00
        intervalFactory.free({ length: new Duration(20 * 60) }, mode),
        // Cooldown: 10:00 30%..30% 85rpm
        intervalFactory.ramp({ length: new Duration(10 * 60), startIntensity: 0.7, endIntensity: 0.3, cadence: 85 }, mode),
      ],
      instructions: [],
    };
    const xml = createWorkoutXml(workout, mode);
    expect(xml).toMatchSnapshot();

    expect(parseWorkoutXml(xml, mode)).toEqual(workout);
  });

  it('creates and parses run workout (duration-based)', () => {
    const mode = createMode({sportType: "run", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});
    const workout: Workout = {
      name: "Run Workout",
      author: "Carl Lewis",
      description: "Featuring a mixture of different interval types.",
      tags: [],
      sportType: mode.sportType,
      lengthType: mode.lengthType,
      intervals: [
        // Warmup: 10:00 30%..75% of marathon pace
        intervalFactory.ramp({ length: new Duration(10 * 60), startIntensity: 0.3, endIntensity: 0.75, pace: PaceType.marathon }, mode),
        // Steady: 5:00 80% of half-marathon pace
        intervalFactory.steady({ length: new Duration(5 * 60), intensity: 0.8, pace: PaceType.halfMarathon }, mode),
        // Intervals: 4 x ON 0:30 120%, OFF 1:00 60% of 5 km pace
        intervalFactory.repetition({ repeat: 4, onLength: new Duration(30), onIntensity: 1.2, offLength: new Duration(60), offIntensity: 0.6, pace: PaceType.fiveKm }, mode),
        // Ramp 1:00 80%..150% of 1 mile pace
        intervalFactory.ramp({ length: new Duration(60), startIntensity: 0.8, endIntensity: 1.5, pace: PaceType.oneMile }, mode),
        // Cooldown: 10:00 30%..30% of 10 km pace
        intervalFactory.ramp({ length: new Duration(10 * 60), startIntensity: 0.7, endIntensity: 0.3, pace: PaceType.tenKm }, mode),
      ],
      instructions: [
        // @ 00:00 - at the start of warmup
        createInstruction({ offset: new Duration(0), text: "Welcome to the workout!" }, mode),
        // @ 12:30 - at the middle of second block
        createInstruction({ offset: new Duration(12 * 60 + 30), text: "This is just a warmup still" }, mode),
        // @ 15:00 - at the start of repetition block
        createInstruction({ offset: new Duration(15 * 60), text: "It's the first one of four sprint efforts" }, mode),
        // @ 19:30 - at the start of last repetition block
        createInstruction({ offset: new Duration(19 * 60 + 30), text: "It's the first one of four sprint efforts" }, mode),
        // @ 21:00 - at the start of a ramp
        createInstruction({ offset: new Duration(21 * 60), text: "As a bonus, we'll ramp up really hard :)" }, mode),
        // @ 31:50 - almost at the very end of cooldown
        createInstruction({ offset: new Duration(31 * 60 + 50), text: "This was it. See you next time." }, mode),
      ],
    };
    const xml = createWorkoutXml(workout, mode);
    expect(xml).toMatchSnapshot();

    expect(parseWorkoutXml(xml, mode)).toEqual(workout);
  });

  it('creates and parses run workout (distance-based)', () => {
    const mode = createMode({sportType: "run", ftp: 200, weight: 75, runningTimes: [], lengthType: "distance"});
    const workout: Workout = {
      name: "Run Workout",
      author: "Carl Lewis",
      description: "Featuring a mixture of different interval types.",
      tags: [],
      sportType: mode.sportType,
      lengthType: mode.lengthType,
      intervals: [
        // Warmup: 1000m 30%..75% of marathon pace
        intervalFactory.ramp({ length: new Distance(1000), startIntensity: 0.3, endIntensity: 0.75, pace: PaceType.marathon }, mode),
        // Steady: 600m 80% of half-marathon pace
        intervalFactory.steady({ length: new Distance(600), intensity: 0.8, pace: PaceType.halfMarathon }, mode),
        // Intervals: 4 x ON 200m 120%, OFF 400m 60% of 5 km pace
        intervalFactory.repetition({ repeat: 4, onLength: new Distance(200), onIntensity: 1.2, offLength: new Distance(400), offIntensity: 0.6, pace: PaceType.fiveKm }, mode),
        // Ramp 600m 80%..150% of 1 mile pace
        intervalFactory.ramp({ length: new Distance(600), startIntensity: 0.8, endIntensity: 1.5, pace: PaceType.oneMile }, mode),
        // Cooldown: 1000m 30%..30% of 10 km pace
        intervalFactory.ramp({ length: new Distance(1000), startIntensity: 0.7, endIntensity: 0.3, pace: PaceType.tenKm }, mode),
      ],
      instructions: [
        // @ 0m - at the start of warmup
        createInstruction({ offset: new Distance(0), text: "Welcome to the workout!" }, mode),
        // @ 1300m - at the middle of second block
        createInstruction({ offset: new Distance(1300), text: "This is just a warmup still" }, mode),
        // @ 1600m - at the start of repetition block
        createInstruction({ offset: new Distance(1600), text: "It's the first one of four sprint efforts" }, mode),
        // @ 3400m - at the start of last repetition block
        createInstruction({ offset: new Distance(3400), text: "It's the first one of four sprint efforts" }, mode),
        // @ 4000m - at the start of a ramp
        createInstruction({ offset: new Distance(4000), text: "As a bonus, we'll ramp up really hard :)" }, mode),
        // @ 5400 - almost at the very end of cooldown
        createInstruction({ offset: new Distance(5400), text: "This was it. See you next time." }, mode),
      ],
    };
    const xml = createWorkoutXml(workout, mode);
    expect(xml).toMatchSnapshot();

    expect(parseWorkoutXml(xml, mode)).toEqual(workout);
  });

  it('excludes instructions past the end of workout', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});
    const workout: Workout = {
      name: "Borderline instruction",
      author: "Who Knows",
      description: "Checking instruction positions.",
      tags: [],
      sportType: mode.sportType,
      lengthType: mode.lengthType,
      intervals: [
        // Steady: 5:00 80%
        intervalFactory.steady({ length: new Duration(5 * 60), intensity: 0.8 }, mode),
      ],
      instructions: [
        // @ 00:00 - at the start of workout
        createInstruction({ offset: new Duration(0), text: "Are we there yet?" }, mode),
        // @ 04:59 - a second before the end of workout
        createInstruction({ offset: new Duration(4 * 60 + 59), text: "Almost there!" }, mode),
        // @ 05:00 - right at the end
        createInstruction({ offset: new Duration(5 * 60), text: "There!" }, mode),
        // @ 08:00 - well past the end
        createInstruction({ offset: new Duration(8 * 60), text: "Enough of it!" }, mode),
      ],
    };
    const xml = createWorkoutXml(workout, mode);
    expect(xml).toMatchSnapshot();

    expect(parseWorkoutXml(xml, mode)).toEqual({ ...workout, instructions: [
      // @ 00:00 - at the start of workout
      createInstruction({ offset: new Duration(0), text: "Are we there yet?" }, mode),
      // @ 04:59 - a second before the end of workout
      createInstruction({ offset: new Duration(4 * 60 + 59), text: "Almost there!" }, mode),
    ]});
  });

  it('ignores comments', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});

    const xml = `
    <workout_file>
      <!-- start the workout -->
      <author>Johnny</author>
      <name>XML with comments</name>
      <description>Featuring a <!-- comment --> inside text.</description>
      <sportType>bike</sportType>
      <durationType>time</durationType>
      <tags>
        <tag name="RECOVERY"/>
        <tag name="INTERVALS"/>
      </tags>
      <workout>
        <SteadyState Duration="300" Power="0.8" pace="0"/>
        <!-- comment between intervals -->
        <SteadyState Duration="300" Power="0.8" pace="0">
          <textevent timeoffset="0" message="Welcome to the interval"/>
          <!-- comment between instructions -->
          <textevent timeoffset="290" message="Near the end now"/>
        </SteadyState>
      </workout>
    </workout_file>`;

    const workout: Workout = {
      name: "XML with comments",
      author: "Johnny",
      description: "Featuring a  inside text.",
      tags: ["RECOVERY", "INTERVALS"],
      sportType: mode.sportType,
      lengthType: mode.lengthType,
      intervals: [
        intervalFactory.steady({ length: new Duration(5 * 60), intensity: 0.8 }, mode),
        intervalFactory.steady({ length: new Duration(5 * 60), intensity: 0.8 }, mode),
      ],
      instructions: [
        createInstruction({ offset: new Duration(5 * 60), text: "Welcome to the interval" }, mode),
        createInstruction({ offset: new Duration(9 * 60 + 50), text: "Near the end now" }, mode),
      ],
    };

    expect(parseWorkoutXml(xml, mode)).toEqual(workout);
  });

  it('throws error when root element is not <workout_file>', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});

    expect(() => parseWorkoutXml("<blah></blah>", mode)).toThrow("Invalid ZWO file");
  });
});
