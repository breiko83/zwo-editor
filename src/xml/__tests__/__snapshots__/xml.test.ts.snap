// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XML creates and parses an empty workout 1`] = `
"<workout_file>
  <author>John Doe</author>
  <name>Sample Workout</name>
  <description>Some intervals for your pleasure (or pain).</description>
  <sportType>bike</sportType>
  <durationType>time</durationType>
  <tags>
    <tag name=\\"RECOVERY\\"/>
    <tag name=\\"INTERVALS\\"/>
  </tags>
  <workout/>
</workout_file>"
`;

exports[`XML creates and parses bike workout (with cadence) 1`] = `
"<workout_file>
  <author>Eddy Merckx</author>
  <name>Cycling Workout</name>
  <description>Featuring a mixture of different interval types.</description>
  <sportType>bike</sportType>
  <durationType>time</durationType>
  <tags/>
  <workout>
    <Warmup Duration=\\"600\\" PowerLow=\\"0.3\\" PowerHigh=\\"0.75\\" pace=\\"0\\" Cadence=\\"80\\"/>
    <SteadyState Duration=\\"300\\" Power=\\"0.8\\" pace=\\"0\\" Cadence=\\"90\\"/>
    <IntervalsT Repeat=\\"4\\" OnDuration=\\"30\\" OffDuration=\\"60\\" OnPower=\\"1.2\\" OffPower=\\"0.6\\" pace=\\"0\\" Cadence=\\"100\\" CadenceResting=\\"60\\"/>
    <Ramp Duration=\\"60\\" PowerLow=\\"0.8\\" PowerHigh=\\"1.5\\" pace=\\"0\\" Cadence=\\"90\\"/>
    <free Duration=\\"1200\\" FlatRoad=\\"0\\"/>
    <Cooldown Duration=\\"600\\" PowerLow=\\"0.7\\" PowerHigh=\\"0.3\\" pace=\\"0\\" Cadence=\\"85\\"/>
  </workout>
</workout_file>"
`;

exports[`XML creates and parses bike workout (without cadence) 1`] = `
"<workout_file>
  <author>Eddy Merckx</author>
  <name>Cycling Workout</name>
  <description>Featuring a mixture of different interval types.</description>
  <sportType>bike</sportType>
  <durationType>time</durationType>
  <tags/>
  <workout>
    <Warmup Duration=\\"600\\" PowerLow=\\"0.3\\" PowerHigh=\\"0.75\\" pace=\\"0\\">
      <textevent timeoffset=\\"0\\" message=\\"Welcome to the workout!\\"/>
    </Warmup>
    <SteadyState Duration=\\"300\\" Power=\\"0.8\\" pace=\\"0\\">
      <textevent timeoffset=\\"150\\" message=\\"This is just a warmup still\\"/>
    </SteadyState>
    <IntervalsT Repeat=\\"4\\" OnDuration=\\"30\\" OffDuration=\\"60\\" OnPower=\\"1.2\\" OffPower=\\"0.6\\" pace=\\"0\\">
      <textevent timeoffset=\\"0\\" message=\\"It's the first one of four sprint efforts\\"/>
      <textevent timeoffset=\\"270\\" message=\\"It's the first one of four sprint efforts\\"/>
    </IntervalsT>
    <Ramp Duration=\\"60\\" PowerLow=\\"0.8\\" PowerHigh=\\"1.5\\" pace=\\"0\\">
      <textevent timeoffset=\\"0\\" message=\\"As a bonus, we'll ramp up really hard :)\\"/>
    </Ramp>
    <free Duration=\\"1200\\" FlatRoad=\\"0\\">
      <textevent timeoffset=\\"600\\" message=\\"Ride as hard as you can for 20 minutes!\\"/>
    </free>
    <Cooldown Duration=\\"600\\" PowerLow=\\"0.7\\" PowerHigh=\\"0.3\\" pace=\\"0\\">
      <textevent timeoffset=\\"590\\" message=\\"This was it. See you next time.\\"/>
    </Cooldown>
  </workout>
</workout_file>"
`;

exports[`XML creates and parses run workout (distance-based) 1`] = `
"<workout_file>
  <author>Carl Lewis</author>
  <name>Run Workout</name>
  <description>Featuring a mixture of different interval types.</description>
  <sportType>run</sportType>
  <durationType>distance</durationType>
  <tags/>
  <workout>
    <Warmup Duration=\\"1000\\" PowerLow=\\"0.3\\" PowerHigh=\\"0.75\\" pace=\\"4\\">
      <textevent timeoffset=\\"0\\" message=\\"Welcome to the workout!\\"/>
    </Warmup>
    <SteadyState Duration=\\"600\\" Power=\\"0.8\\" pace=\\"3\\">
      <textevent timeoffset=\\"300\\" message=\\"This is just a warmup still\\"/>
    </SteadyState>
    <IntervalsT Repeat=\\"4\\" OnDuration=\\"200\\" OffDuration=\\"400\\" OnPower=\\"1.2\\" OffPower=\\"0.6\\" pace=\\"1\\">
      <textevent timeoffset=\\"0\\" message=\\"It's the first one of four sprint efforts\\"/>
      <textevent timeoffset=\\"1800\\" message=\\"It's the first one of four sprint efforts\\"/>
    </IntervalsT>
    <Ramp Duration=\\"600\\" PowerLow=\\"0.8\\" PowerHigh=\\"1.5\\" pace=\\"0\\">
      <textevent timeoffset=\\"0\\" message=\\"As a bonus, we'll ramp up really hard :)\\"/>
    </Ramp>
    <Cooldown Duration=\\"1000\\" PowerLow=\\"0.7\\" PowerHigh=\\"0.3\\" pace=\\"2\\">
      <textevent timeoffset=\\"800\\" message=\\"This was it. See you next time.\\"/>
    </Cooldown>
  </workout>
</workout_file>"
`;

exports[`XML creates and parses run workout (duration-based) 1`] = `
"<workout_file>
  <author>Carl Lewis</author>
  <name>Run Workout</name>
  <description>Featuring a mixture of different interval types.</description>
  <sportType>run</sportType>
  <durationType>time</durationType>
  <tags/>
  <workout>
    <Warmup Duration=\\"600\\" PowerLow=\\"0.3\\" PowerHigh=\\"0.75\\" pace=\\"4\\">
      <textevent timeoffset=\\"0\\" message=\\"Welcome to the workout!\\"/>
    </Warmup>
    <SteadyState Duration=\\"300\\" Power=\\"0.8\\" pace=\\"3\\">
      <textevent timeoffset=\\"150\\" message=\\"This is just a warmup still\\"/>
    </SteadyState>
    <IntervalsT Repeat=\\"4\\" OnDuration=\\"30\\" OffDuration=\\"60\\" OnPower=\\"1.2\\" OffPower=\\"0.6\\" pace=\\"1\\">
      <textevent timeoffset=\\"0\\" message=\\"It's the first one of four sprint efforts\\"/>
      <textevent timeoffset=\\"270\\" message=\\"It's the first one of four sprint efforts\\"/>
    </IntervalsT>
    <Ramp Duration=\\"60\\" PowerLow=\\"0.8\\" PowerHigh=\\"1.5\\" pace=\\"0\\">
      <textevent timeoffset=\\"0\\" message=\\"As a bonus, we'll ramp up really hard :)\\"/>
    </Ramp>
    <Cooldown Duration=\\"600\\" PowerLow=\\"0.7\\" PowerHigh=\\"0.3\\" pace=\\"2\\">
      <textevent timeoffset=\\"590\\" message=\\"This was it. See you next time.\\"/>
    </Cooldown>
  </workout>
</workout_file>"
`;

exports[`XML excludes instructions past the end of workout 1`] = `
"<workout_file>
  <author>Who Knows</author>
  <name>Borderline instruction</name>
  <description>Checking instruction positions.</description>
  <sportType>bike</sportType>
  <durationType>time</durationType>
  <tags/>
  <workout>
    <SteadyState Duration=\\"300\\" Power=\\"0.8\\" pace=\\"0\\">
      <textevent timeoffset=\\"0\\" message=\\"Are we there yet?\\"/>
      <textevent timeoffset=\\"299\\" message=\\"Almost there!\\"/>
    </SteadyState>
  </workout>
</workout_file>"
`;
